File: full_case_study_solns/_e_ballbeam/matlab_old/hw10/ballbeamSim.m
restoredefaultpath;
ballbeamParamHW10;  % load parameters

% instantiate ballbeam, controller, and reference input classes 
% Instantiate Dynamics class
ballbeam = ballbeamDynamics(P);  
ctrl = ballbeamController(P);  
    amplitude = 0.125; % amplitude of reference input
    frequency = 0.02; % frequency of reference input
reference = signalGenerator(amplitude, frequency);  

% instantiate the data plots and animation
dataPlot = plotData(P);
animation = ballbeamAnimation(P);

disturbance = 1.0;

% main simulation loop
t = P.t_start;  % time starts at t_start
while t < P.t_end  
    % Get referenced inputs from signal generators
    ref_input = 0.25 + reference.square(t);
    % Propagate dynamics in between plot samples
    t_next_plot = t + P.t_plot;
    while t < t_next_plot % updates control and dynamics at faster simulation rate
        u = ctrl.u(ref_input, ballbeam.outputs());  % Calculate the control value
        input = u + disturbance;
        ballbeam.propagateDynamics(input);  % Propagate the dynamics
        t = t + P.Ts; % advance time by Ts
    end
    % update animation and data plots
    animation.drawBallbeam(ballbeam.states);
    dataPlot.updatePlots(t, ref_input, ballbeam.states, u);
end

File: full_case_study_solns/_e_ballbeam/python_old/hw18/ballbeamSim.py
import sys
sys.path.append('..')  # add parent directory
import matplotlib.pyplot as plt
import ballbeamParam as P
from ballbeamDynamics import ballbeamDynamics
from ballbeamController import ballbeamController
from signalGenerator import signalGenerator
from ballbeamAnimation import ballbeamAnimation
from plotData import plotData

# instantiate ballbeam, controller, and reference classes
ballbeam = ballbeamDynamics()
ctrl = ballbeamController()
reference = signalGenerator(amplitude=0.125, frequency=0.04)

# instantiate the simulation plots and animation
dataPlot = plotData()
animation = ballbeamAnimation()

# set disturbance input
disturbance = 1.0

t = P.t_start  # time starts at t_start
while t < P.t_end:  # main simulation loop
    # Get referenced inputs from signal generators
    ref_input = reference.square(t)
    ref_input[0] += 0.25
    # Propagate dynamics in between plot samples
    t_next_plot = t + P.t_plot
    while t < t_next_plot: # updates control and dynamics at faster simulation rate
        u = ctrl.u(ref_input, ballbeam.outputs())  # Calculate the control value
        input = [u[0]+disturbance]
        ballbeam.propagateDynamics(input)  # Propagate the dynamics
        t = t + P.Ts  # advance time by Ts
    # update animation and data plots
    animation.drawBallbeam(ballbeam.states())
    dataPlot.updatePlots(t, ref_input, ballbeam.states(), u)
    plt.pause(0.0001)  # the pause causes the figure to be displayed during the simulation

# Keeps the program from closing until the user presses a button.
print('Press key to close')
plt.waitforbuttonpress()
plt.close()
